src/main/java/edu/kit/compiler/semantic/SymbolTable.java

```diff
@@ -0,0 +1,143 @@
+package edu.kit.compiler.semantic;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import java.util.ArrayDeque;
+import java.util.LinkedHashMap;
+
+@Data
+public class SymbolTable  {
+
+    private final ArrayDeque<Change> changes = new ArrayDeque<>();
+    private final SymbolStringTable symbolStringTable = new SymbolStringTable();
+    private Scope currentScope;
+
+    @RequiredArgsConstructor
+    static class Symbol {
+        @NonNull
+        int name;
+        Scope currentScope;
+        Definition currentDefinition;
+    }
+
+    @AllArgsConstructor
+    static class Scope {
+        Scope parent;
+        int oldSize;
+    }
+
+    @AllArgsConstructor
+    static class Change {
+        Symbol sym;
+        Definition previousDefinition;
+        Scope previousScope;
+
+    }
+
+    /**
+     * AstNodes do not hold a String reference, they only hold a reference integer to our global string table.
+     * Thus the mapping from String to symbol can be simplified by only providing a mapping from
+     * Integer -> Symbol instead of String -> Integer -> Symbol
+     */
+    static class SymbolStringTable {
+
+        private final LinkedHashMap<Integer, Symbol> table = new LinkedHashMap<>();
+
+        /**
+         *
+         * @param name The name of the symbol mapped by our global String table
+         * @return The symbol for which this name is mapped otherwise null
+         */
+        public Symbol find(int name) {
+            return table.get(name);
+        }
+        /**
+         * Returns a symbol that has yet to be initialized with current scope and current definition
+         * @param name The name of the symbol
+         * @return If there exists a symbol in the string table return the symbol otherwise create a new symbol
+         */
+        public Symbol insert(int name) {
+            if(table.get(name) == null) {
+                Symbol s = new Symbol(name);
+                table.put(name, s);
+                return s;
+            }
+            else {
+                return table.get(name);
+            }
+        }
+
+        public void remove(int name) {
+            table.remove(name);
+        }
+    }
+
+    public void enterScope() {
+        currentScope = new Scope(currentScope, changes.size());
+    }
+    public void leaveScope() {
+        // revert all previous changes
+        while (changes.size() > currentScope.oldSize) {
+            Change c = changes.pop();
+            // check if we leave our defining scope then remove symbol entirely from table
+            if (c.previousScope == null) {
+                symbolStringTable.remove(c.sym.name);
+                return;
+            }
+            c.sym.currentDefinition = c.previousDefinition;
+            c.sym.currentScope = c.previousScope;
+        }
+        currentScope = currentScope.parent;
+    }
+    public Symbol insert(int name, Definition definition) {
```

Comment:
Why is the additional `name` parameter necessary? Is there a scenario where we want to insert a definition with a name different from `definition.getName()`?

Reply:
Right now I can't think of one. I mean we could leave it out and when we come across a case where this is necessary we add it again.